### ISSUE TITLE
Poll MUSH via HTTP for info

### ISSUE THREAD

**Issue by talvo** [pennmush/pennmush#1093]
Allow polling the MUSH for info via an HTTP GET request. (Allowing POST would be possible too, but since the info doesn't all come in on the first line then, it requires more special handling, so I suggest not bothering.)

Easiest format would be to have it all in the URL: /json/who/etc/ to return info on connected players (who) and something else (etc)

A better, but not quite so easy, format would be to use proper args for it: /json/?who=true&etc=true for the same info (json: {"who": true, "etc": true} )

Could also be made more complex by, for instance, /json/?who[]=name&who[]=idle&etc=true to filter the info to only return the names and idle times of connected players instead of all info for "who" (json: {"who": ["name", "idle"], "etc": true} )

Information available could be limited to a small number of predefined items ('who' for connected player details, 'mssp' for the defined MSSP data, maybe 'config' for \@config stuff? Anything else useful?) or there could be a config option pointing to a handler attr used for retrieving custom, MUSH-specific fields.

If I had any other thoughts I've already forgotten them, which is why I'm making this ticket instead of writing the code now. \:p  Thoughts welcome.

**Comment by tkrajcar** [comment]
Combining multiple pieces of disparate info in one request breaks REST modeling; would prefer to avoid it if possible.

So, proposed URLs would look more like:
/who.json
/who.json?fields=name,idle

**Comment by tkrajcar** [comment]
Oh, and, I'm a huge 👍 on this idea.

**Comment by talvo** [comment]
I would argue it's not really disparate as it's all info on the current state of the MUSH. It also means the MUSH isn't receiving dozens of connections for this when one would suffice. But if there are big complaints about that then that's fair enough.

Instead of /who.json, though, I'd still prefer /json/who for simple parsing reasons. And also ?fields[]=name&fields[]=idle for nicer building of the requests via jQuery/javascript/whatever.

**Comment by HarryCordewener** [comment]
Reminds me of 
https://github.com/pennmush/pennmush/issues/1032

:)

**Comment by captdeaf** [comment]
I think we should present it as a unified softcode interface, with hardcoded "defaults":

1) JSON attribute call.

Syntax should be GET /dbref/attrname, with all arguments passed as a json object in %0, is equivalent to `u(#dbref/attrname, %0)`

It's easy enough to convert get parameters into json. `GET /...?fields[]=who&fields[]=info` can be converted to `{"fields": ["info", "who"]}`

GET /dbref alone calls `u(#dbref/GET, %0)`

Permissions: An object must be `@power`ed HTTP, and the attribute called must have http_get attrflag. (So helper functions can't be called directly). GET must exist and have http_get attrflag in order to be called, too, otherwise `/dbref` should 404.

2) `json_map_object`:
Add a sister to json_map - `json_map_object`, which, instead of returning a single string, returns a js on object, and can only take a json object.

e.g: `json_map_object(#apply/reverse, \{"foo": "olleh"\, "bar": "blah"\})` returns `{"foo": "hello", "bar": "halb"}`

3) Hardcoded INFO/WHO/etc handler

Add an \@config value of "http_info_handler", much like our event handler. If that doesn't exist, then:

GET /json is treated as a fake object (real if http_info_handler exists) which is basically equivalent to:
```
&GET object=json_map_object(#apply2/u, %0)
&WHO object=...
&INFO object=...
&etc...
```
(And all set http).

4) POST and @-commands.

POST /dbref/attr - Object must be `@powered` http, and attr have http_post flag.

Should be equivalent of running `@include dbref/attr`, and redirecting the `@pemit`s to the socket. Any queued commands won't work, naturally, everything would need to be `@include`'d

**Comment by HarryCordewener** [comment]
I recommend a 'default REST endpoint' able to be set up in the Config as well, for centralization of function calls etc. 

Ex: GET /REST/attrname

**Comment by captdeaf** [comment]
Yeah, I think that should be @config http_info_handler. @talvo seems to want it /json/ instead of /rest/, but I'm fine with anything.

**Comment by talvo** [comment]
I personally think that having a bunch of powers and attribute flags for this is overkill, and that mapping it directly to attributes on objects probably isn't desirable. To my mind, having one attribute that gets called for everything is a lot more desirable.

You can still achieve what you're describing really easily with that:
```
&whatever obj=switch(%0,#*/*,u(%0,%1),...)
GET /REST/#123/foo {"your":"json here"}
```
but without all the extra complication. 

(Also, just for the record, 'easy enough'? Converting them was a bitch.  :wink:)

**Comment by talvo** [comment]
Oh, re 4 - any particular reason for allowing commands for POST? Also seems like an incredible complication to add for very little gain.

**Comment by captdeaf** [comment]
"was" a bitch, so it's easy enough to do it now? :smirk:

POST bit is a nice to have, can be completely separate from GET. But it should follow our paradigm - functions get things, commands do things. GET gets things, POST does things.

**Comment by talvo** [comment]
Generally with a REST interface, using POST just means you're trying to send something that's too long to send via GET.

**Comment by tkrajcar** [comment]
Forewarning: I spend all day dealing with other people's APIs, so my opinions about these things tend to be rather pointy.

A few broad points I'd like to ask be considered:
1. POST should only be used for 'write' operations, and GET only for 'read' operations; to say it another way, a client running the same GET operation twice (assuming no other changes to the underlying system) should receive identical output and cause no system changes.
2. `/foo.json` is vastly preferable to `/json/foo` -- even if we never plan to have any other serialization methods. Or, if we're extremely confident it's JSON for life, omit it and just have `/foo`.
3. I would suggest an iterative (not to be confused with iter()ative) implementation where we start with a fairly limited/basic feature set, but solve all the HTTP server piping issues. There's plenty of time to implement the kitchen-sink of features, but it'd be nice to get something out there that people can start testing and working the bugs out of.
4. This raises a number of security concerns, including concerns like limiting POST size, defeating slow-request DoS attacks; are we proposing that users would expose a HTTP(S...?) server written into PennMUSH to the outside directly, or is it intended that everyone will have a nginx (or whatever your favorite is) front-end proxy in place to proxy these requests and deal with some of these issues? Or, are we expecting users to only access this endpoint from trusted sources and use firewalling to block all others? Just curious what thoughts are out there.

**Comment by talvo** [comment]
@tkrajcar, that's a little ahead of where we're at right now. All I'm currently working on is being able to connect on the normal MUSH port, send GET /rest/something.json?args=foo (which seems to be the most agreed-upon format) and get JSON back. That's pretty much the limit of what I currently have planned, assuming we ever manage to agree on how to handle it in-MUSH from there. No more risk from DoS than with any other connection to the MUSH with that.

**Comment by grapenut** [comment]
Websocket support solves this problem without the need to implement any sort of query parser/http server, is automatically extensible in the sense that people wouldn't have to hack any additional hardcode to add new commands/features, and employs all the same security features as are used already in the game. Open a websocket to penn, use a special webguest player to execute some softcode or run some preset commands and read the results. The information can even be passed back as a JSON object.

Also it lets people connect to mush from easy to make webclients.
